livekit-agents/livekit/agents/llm/llm.py

```diff
@@ -29,35 +29,33 @@ class LLM(abc.ABC):
     @abc.abstractmethod
     async def chat(
         self,
-        history: ChatContext,
-        fnc_ctx: FunctionContext | None = None,
+        *,
+        chat_ctx: ChatContext,
+        fnc_ctx: function_context.FunctionContext | None = None,
         temperature: float | None = None,
         n: int | None = None,
     ) -> "LLMStream": ...
 
 
-@dataclass
-class CalledFunction:
-    fnc_name: str
-    fnc: Callable
-    args: dict
-
-
 class LLMStream(abc.ABC):
     def __init__(self) -> None:
-        # fnc_name, args..
-        self._called_functions: list[CalledFunction] = []
+        self._called_functions: list[function_context.CalledFunction] = []
 
     @property
-    def called_functions(self) -> list[CalledFunction]:
+    def called_functions(self) -> list[function_context.CalledFunction]:
         """List of called functions from this stream."""
         return self._called_functions
 
     @abc.abstractmethod
-    def __aiter__(self) -> AsyncIterator[ChatChunk]: ...
+    async def gather_function_results(
+        self,
+    ) -> list[function_context.CalledFunction]: ...
+
+    def __aiter__(self) -> AsyncIterator[ChatChunk]:
+        return self
 
     @abc.abstractmethod
     async def __anext__(self) -> ChatChunk: ...
 
     @abc.abstractmethod
-    async def aclose(self, wait: bool = True) -> None: ...
+    async def aclose(self) -> None: ...
```

Comment:
how do we ensure that the reader has consumed all of the content then?

Reply:
we never have to for LLM, you push once. Also when the stream ends, you know you have consumed everything.

There is also the async function gather_function_results that is used to make sure we wait for all function calls to resolve

Reply:
Also I'll make a PR and a changelog for it, but I have plans to remove this argument for every baseclass we have.
I think it is better to have an explicit function in most cases.